bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/api/PredefinedTypes.java

```diff
@@ -164,9 +164,10 @@
     public static final ServiceType TYPE_ANY_SERVICE = new BServiceType(TypeConstants.SERVICE, EMPTY_MODULE, 0);
     public static final HandleType TYPE_HANDLE = new BHandleType(TypeConstants.HANDLE_TNAME, EMPTY_MODULE);
     public static final UnionType ANYDATA_OR_READONLY = new BUnionType(Arrays.asList(TYPE_ANYDATA, TYPE_READONLY));
-    public static final MapType TYPE_ERROR_DETAIL = new BMapType(TypeConstants.MAP_TNAME, ANYDATA_OR_READONLY,
-                                                                 EMPTY_MODULE);
-    public static final ErrorType TYPE_ERROR = new BErrorType(TypeConstants.ERROR, EMPTY_MODULE, TYPE_ERROR_DETAIL);
+    public static final Type TYPE_ERROR_DETAIL = ReadOnlyUtils.setImmutableTypeAndGetEffectiveType(
+                                            new BMapType(TypeConstants.MAP_TNAME, ANYDATA_OR_READONLY, EMPTY_MODULE));
+    public static final ErrorType TYPE_ERROR = new BErrorType(TypeConstants.ERROR, EMPTY_MODULE,
+                                            new BMapType(TypeConstants.MAP_TNAME, ANYDATA_OR_READONLY, EMPTY_MODULE));
```

Comment:
Can't we use `TYPE_ERROR_DETAIL` here? This also has to be immutable, right?

Reply:
In here we set error values as immutable. So if we change `error` type as immutable we can't do below kind of casting.
```ballerina
type CustomErrorData record {
    string message?;
    error cause?;
    int accountID?;
};
type CustomError1 distinct error<CustomErrorData>;
type CustomError distinct error<CustomErrorData>;

type Error CustomError | CustomError1;

class Foo {
    public isolated function next() returns record {| int value; |}|Error? {

        CustomError e = CustomError("CustomError", message = "custom error occured", accountID = 2);
        return e;
        
    }
}

public function main() {
    Foo val = new;
    object {
        public isolated function next() returns record {|int value;|}|error?;} err = <object {
         public isolated function next() returns record {|int value;|}|error?;}> val;
}
```

Reply:
All error values/types and their detail values/types have to be immutable. If using the immutable detail here causes something to fail, it's possible that we're creating an error with a non-immutable detail somewhere, which is wrong. Can you check why exactly it fails?

Reply:
When creating `CustomError1` we are not set detail(`CustomErrorData`) as immutable. Is it effect when creating values from that type( `CustomErrorData`) If we set it to `immutable`.

Reply:
That's wrong, right? When the error value is being created, cloneReadOnly should get called for the detail, so at runtime the detail type should be a subtype of `readonly`?

Reply:
As discussed offline, let's have the mutable detail type here for the time-being because that's how we've modeled the errors defined in the source too.